torch/nn/parameter.py

```diff
@@ -44,3 +44,88 @@ def __reduce_ex__(self, proto):
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration. Moreover, Uninitialized parameters
+    become full fledged :class:`torch.nn.Parameter` after they are initialized.
+    """
+    def __new__(cls, requires_grad=True):
+        data = torch.Tensor()
+        return torch.Tensor._make_subclass(cls, data, requires_grad)
+
+    def materialize(self, shape, device=None, dtype=None):
+        r"""Create a Parameter with the same properties of the uninitialized one.
+        Given a shape, it materializes a parameter in the same device
+        and with the same `dtype` as the current one or the specified ones in the 
+        arguments
+        Args:
+            shape : (tuple): the shape for the materialized tensor.
+            device (:class:`torch.device`): the desired device of the parameters
+                and buffers in this module. Optional.
+            dtype (:class:`torch.dtype`): the desired floating point type of
+                the floating point parameters and buffers in this module. Optional.
+        """
+        if device is None:
+            device = self.data.device
+        if dtype is None:
+            dtype = self.data.dtype
+        self.data = torch.empty(shape, device=device, dtype=dtype)
+        self.__class__ = Parameter
+
+    @property
+    def shape(self):
+        raise RuntimeError(
+            'Can\'t access the shape of an uninitialized parameter. '
+            'This error usually happens in `load_state_dict` when trying to load '
+            'an uninitialized parameter into an initialized one. '
+            'Call `forward` to initialize the parameters before accessing their attributes.')
+
+    def share_memory_(self):
+        raise RuntimeError(
+            'Can\'t share memory on an uninitialized parameter. '
+            'Call `forward` to initialize the parameters before calling '
+            '`module.share_memory()`.')
+
+    def __repr__(self):
+        return 'Uninitialized parameter'
+
+    def __reduce_ex__(self, proto):
+        # See Note [Don't serialize hooks]
+        return (
+            UninitializedParameter,
+            (self.requires_grad,)
+        )
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        allowed_methods = [
+            torch.Tensor.__hash__,
+            torch.Tensor.size,
+            torch.Tensor.copy_,
+            torch.Tensor.is_floating_point,
+            torch.Tensor.half,
+            torch.Tensor.float,
+            torch.Tensor.double,
+            torch.Tensor.char,
+            torch.Tensor.short,
+            torch.Tensor.int,
+            torch.Tensor.long,
+            torch.Tensor.cuda,
+            torch.Tensor.cpu,
+            torch.Tensor.to,
+            torch._has_compatible_shallow_copy_type]
+        # method-wrapper is to detect access to Tensor properties that are 
+        # wrapped in descriptors
```

Comment:
what exactly does it mean? Is it stuff like `torch.shape`?

Reply:
when calling an attribute that is accessed through a [descriptor](https://docs.python.org/3/howto/descriptor.html) the func object that we get is a bit weird